UI/ChatWnd.cpp

```diff
@@ -611,42 +613,104 @@ void MessageWnd::OpenForInput() {
 
 namespace {
     void SendChatMessage(const std::string& text, std::set<int> recipients) {
+        const ClientApp* app = ClientApp::GetApp();
+        if (!app) {
+            ErrorLogger() << "ChatWnd.cpp SendChatMessage couldn't get client app!";
+            return;
+        }
         ClientNetworking& net = HumanClientApp::GetApp()->Networking();
+        auto players = app->Players();
+        std::string recipient_names = "";
 
         if (recipients.empty()) {
             net.SendMessage(PlayerChatMessage(text));
         } else {
-            recipients.insert(HumanClientApp::GetApp()->PlayerID());   // ensure recipient sees own sent message
-            for (int recipient_id : recipients)
-                net.SendMessage(PlayerChatMessage(text, recipient_id));
+            for (int recipient_id : recipients) {
+                net.SendMessage(PlayerChatMessage(UserString("MESSAGES_WHISPER") + " " + text, recipient_id));
+
+                auto player_it = players.find(recipient_id);
+                if (player_it != players.end())
+                    recipient_names += player_it->second.name + "; ";
+            }
+            recipient_names = recipient_names.substr(0, recipient_names.size() - 2);
+            net.SendMessage(PlayerChatMessage("[" + UserString("TO") + ": " + recipient_names + + "] " + text, HumanClientApp::GetApp()->PlayerID()));
```

Comment:
Same here... the message metadata shouldn't be baked in on the sender's side

Reply:
Should be fixed.